MetaMask/metamask-extension
# Fix sentry sessions

## **Description**

The session count in our Sentry dashboard has decreased drastically since v11.1.0.

It seems that the only time the session count is incremented is when using `captureException` and `captureMessage` methods. This explains why the session count is much lower than previously but not zero. It also explains a crash rate close to but different than 100% — because `captureMessage` is much less used than captureException on our repository.

### Investigation Notes & Fix

The `toggleSession()` function call inside `setParticipateInMetaMetrics()` in the `ui/store/actions.ts` file is redundant, as both `onboarding/metametrics.js` (for onboarding) and `ui/pages/settings/security-tab/security-tab.component.js` (for the toggling flow) call `setParticipateInMetaMetrics()` on the MetaMetrics controller that starts or stops the session appropriately and updates the state directly. Moreover, by the time `toggleSession()` is called inside `setParticipateInMetaMetrics()` in the `ui/store/actions.ts` file, the `participateInMetaMetrics` property in the MetaMetrics controller state has not yet been updated. After removing that `toggleSession()` function call, we can see that the state is changed on both flows by using `copy(await window.stateHooks.getCleanAppState())`.

We also don’t need to use `getMetaMetricsEnabled()` for `startSession()` and `endSession()` anymore because `setParticipateInMetaMetrics()` is explicitly called with an argument that dictates whether to call the start session or end session methods. As such, I removed `getMetaMetricsEnabled` from the conditions on `startSession` and `endSession`.

Using `Hub.captureSession` and `Hub.captureSession(true)` on start and end session respectively yields the right behaviour for the `https://sentry.io/api/<project_id>/envelope/?sentry_key=<project secret>` requests. When starting a session the request payload includes `"init": true` and when ending a session, the request payload includes `"init": false` and a `"duration"` property.

Finally, the sessions are still not captured on the sentry dashboard, unless the `@sentry/browser` package is updated. This may be because we used the newer version for testing with [`sentry-example`](https://github.com/MetaMask/sentry-example/), and two different package versions on the same project may confuse the dashboard internals. Upgrading the package to the latest version fixes this.

### Summary of changes:
- Remove redundant `toggleSession()` function call from `setParticipateInMetaMetrics` in `ui/store/actions.ts`.
- Remove `toggleSession()` function definition from `app/scripts/lib/setupSentry.js` because it’s not used anymore.
- Remove usage of `getMetaMetricsEnabled()` inside `startSession()` and `endSession()` in `app/scripts/lib/setupSentry.js` because repeated calls for both methods are not possible.
- Update sentry browser package version from ^7.53.0 to ^7.81.1 so the sessions are logged on the sentry dashboard.


## **Manual testing steps**

1. Create a new project on sentry as per the [documentation](https://github.com/MetaMask/metamask-extension/blob/1195f7c2fa7ee7b43db2fa62dcb85e26300bbbb0/README.md?plain=1#L24)
2. Remove the extension completely
3. Build the extension using `yarn dist`
4. Install the extension from the `dist` folder
5. Open the dev tools on the network tab for the UI and background processes.
6. Onboard opting into MetaMetrics
7. See the Sentry request with type “session” and “init” true.
8. You should now see the sessions on Releases > Adoption.

7. Reload the Extension on the Extension tab
9. Repeat Steps 5-8

10. Reload the Extension UI in the browser
11. Repeat Steps 5-8

12. When opting in and out of metametrics, Sentry session requests should be visible as well.

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
